File: server/Web/src/Core/Domain/Domain/Enums/Currency.cs
using Domain.Core.Primitives;

namespace Domain.Enums;

public sealed class Currency : Enumeration<Currency>
{
    public static readonly Currency Usd = new(1, "Dollar", "USD");
    public static readonly Currency Eur = new(2, "Euro", "EUR");
    public static readonly Currency Pln = new(3, "Złoty", "PLN");

    internal static readonly Currency None = new(0, string.Empty, string.Empty);

    private Currency(int value, string name, string code)
        : base(value, name)
    {
        Code = code;
    }

    // Ctor for EF
    // ReSharper disable once UnusedMember.Local
    private Currency()
    {
        Code = string.Empty;
    }

    public string Code { get; }

    public string Format(decimal amount) => $"{amount:n2} {Code}";
}

File: client/src/app/core/enums/Currency.ts
export class Currency {
	static USD = {
		id: 1,
		name: 'USD',
	};
	static EUR = {
		id: 2,
		name: 'EUR',
	};
	static PLN = {
		id: 3,
		name: 'PLN',
	};

	static fromName(name: string): any {
		switch (name) {
			case 'USD':
				return Currency.USD;
			case 'EUR':
				return Currency.EUR;
			case 'PLN':
				return Currency.PLN;
			default:
				throw new Error(`Unknown currency: ${name}`);
		}
	}

	static fromId(id: number): any {
		switch (id) {
			case Currency.USD.id:
				return Currency.USD;
			case Currency.EUR.id:
				return Currency.EUR;
			case Currency.PLN.id:
				return Currency.PLN;
			default:
				throw new Error(`Unknown currency id: ${id}`);
		}
	}
}
